// @valaxyjs/devtools@0.19.9/dist/index.cjs
'use strict';

const c = require('picocolors');
const sirv = require('sirv');
const httpProxyMiddleware = require('http-proxy-middleware');
const node_path = require('node:path');
const node_url = require('node:url');
const bodyParser = require('body-parser');
const matter = require('gray-matter');
const jsYaml = require('js-yaml');
const fs = require('fs-extra');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const c__default = /*#__PURE__*/_interopDefaultCompat(c);
const sirv__default = /*#__PURE__*/_interopDefaultCompat(sirv);
const bodyParser__default = /*#__PURE__*/_interopDefaultCompat(bodyParser);
const matter__default = /*#__PURE__*/_interopDefaultCompat(matter);
const fs__default = /*#__PURE__*/_interopDefaultCompat(fs);

const DIR_DIST = typeof __dirname !== "undefined" ? __dirname : node_path.dirname(node_url.fileURLToPath((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href))));
const DEVTOOLS_CLIENT_FOLDER = node_path.resolve(DIR_DIST, "../dist/client");
const DIR_CLIENT = DEVTOOLS_CLIENT_FOLDER;

const prefix = "/valaxy-devtools-api";
async function migration(path, frontmatter) {
  if (fs__default.existsSync(path)) {
    const rawMd = await fs__default.readFile(path, "utf-8");
    const matterFile = matter__default(rawMd, { schema: jsYaml.JSON_SCHEMA });
    let mod = false;
    for (const key in frontmatter) {
      if (key in matterFile.data) {
        matterFile.data[frontmatter[key]] = matterFile.data[key];
        delete matterFile.data[key];
        mod = true;
      }
    }
    if (mod) {
      const newMd = matter__default.stringify(matterFile.content, matterFile.data);
      await fs__default.writeFile(path, newMd);
    }
  }
}
function registerApi(server, _viteConfig) {
  const app = server.middlewares;
  app.use(bodyParser__default.json());
  app.use(`${prefix}/frontmatter`, async (req, _res) => {
    if (req.method === "POST") {
      const { pageData, frontmatter: newFm } = await req.body;
      const path = pageData.path;
      if (fs__default.existsSync(path)) {
        const rawMd = await fs__default.readFile(path, "utf-8");
        const matterFile = matter__default(rawMd);
        matterFile.data = newFm;
        const newMd = matter__default.stringify(matterFile.content, matterFile.data);
        await fs__default.writeFile(path, newMd);
      }
    }
  });
  app.use(`${prefix}/migration`, async (req, _res) => {
    if (req.method === "POST") {
      const { pageData, frontmatter } = await req.body;
      const worker = [];
      for (const item of pageData) {
        const path = item;
        worker.push(migration(path, frontmatter));
      }
      Promise.all(worker).then(() => {
        _res.end("ok");
      }).catch((_) => {
        _res.end("migration error");
      });
    }
  });
}

const NAME = "valaxy:devtools";
function ValaxyDevtools(options = {}) {
  let config;
  function configureServer(server) {
    const _print = server.printUrls;
    const base = (options.base ?? server.config.base) || "/";
    const devtoolsUrl = `${base}__valaxy_devtools__/`;
    if (undefined?.VITE_DEV_VALAXY_DEVTOOLS === "true") {
      server.middlewares.use(devtoolsUrl, httpProxyMiddleware.createProxyMiddleware({
        target: "http://localhost:5001/#/",
        changeOrigin: true
      }));
    } else {
      server.middlewares.use(devtoolsUrl, sirv__default(DIR_CLIENT, {
        single: true,
        dev: true
      }));
    }
    server.printUrls = () => {
      let host = `${config.server.https ? "https" : "http"}://localhost:${config.server.port || "80"}`;
      const url = server.resolvedUrls?.local[0];
      if (url) {
        try {
          const u = new URL(url);
          host = `${u.protocol}//${u.host}`;
        } catch (error) {
          console.warn("Parse resolved url failed:", error);
        }
      }
      _print();
      const colorUrl = (url2) => c__default.green(url2.replace(/:(\d+)\//, (_, port) => `:${c__default.bold(port)}/`));
      console.log(`  ${c__default.green("\u279C")}  ${c__default.bold("Inspect")}: ${colorUrl(`${host}${base}__inspect/`)}`);
    };
    registerApi(server);
  }
  const plugin = {
    name: NAME,
    enforce: "pre",
    configResolved(_config) {
      config = _config;
    },
    configureServer(server) {
      configureServer(server);
    }
  };
  return plugin;
}

module.exports = ValaxyDevtools;
